velox/expression/CastExpr.cpp

```diff
@@ -602,99 +619,59 @@ void CastExpr::apply(
-
-      switch (toType->kind()) {
-        // Handle complex type conversions
-        case TypeKind::MAP:
-          localResult = applyMap(
-              *translatedRows,
-              decoded->base()->asUnchecked<MapVector>(),
-              context,
-              fromType->asMap(),
-              toType->asMap());
-          break;
-        case TypeKind::ARRAY:
-          localResult = applyArray(
-              *translatedRows,
-              decoded->base()->asUnchecked<ArrayVector>(),
-              context,
-              fromType->asArray(),
-              toType->asArray());
-          break;
-        case TypeKind::ROW:
-          localResult = applyRow(
-              *translatedRows,
-              decoded->base()->asUnchecked<RowVector>(),
-              context,
-              fromType->asRow(),
-              toType->asRow());
-          break;
-        case TypeKind::SHORT_DECIMAL:
-        case TypeKind::LONG_DECIMAL:
-          localResult = applyDecimal(
-              *translatedRows, *decoded, context, fromType, toType);
-          break;
-        default: {
-          VELOX_UNREACHABLE();
-        }
-      }
+  if (castFromOperator_ || castToOperator_ || !toType->isPrimitiveType() ||
+      isDecimalKind(toType->kind())) {
+    VectorPtr localResult;
 
-      if (!decoded->isIdentityMapping()) {
-        localResult = decoded->wrap(localResult, *input, rows);
+    if (decoded->isIdentityMapping()) {
+      applyComplexAndCustomType(
+          *nonNullRows, *decoded, context, fromType, toType, localResult);
+    } else {
+      ContextSaver saver;
+      LocalSelectivityVector translatedRowsHolder(*context.execCtx());
+
+      if (decoded->isConstantMapping()) {
+        setConstantWrappingAndTranslateRows(
+            decoded->index(nonNullRows->begin()),
+            *nonNullRows,
+            context,
+            saver,
+            translatedRowsHolder);
+      } else {
+        setDictionaryWrappingAndTranslateRows(
+            *decoded,
+            input,
+            *nonNullRows,
+            context,
+            saver,
+            translatedRowsHolder);
       }
 
-      context.moveOrCopyResult(localResult, rows, result);
-      context.releaseVector(localResult);
-    } else {
-      // Handling primitive type conversions
-      context.ensureWritable(rows, toType, result);
-      // Unwrapping toType pointer. VERY IMPORTANT: dynamic type pointer and
-      // static type templates in each cast must match exactly
-      VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
-          applyCast,
-          toType->kind(),
-          fromType->kind(),
-          *nonNullRows,
-          context,
+      applyComplexAndCustomType(
+          *translatedRowsHolder,
           *decoded,
-          result);
+          context,
+          fromType,
+          toType,
+          localResult);
+
+      localResult = decoded->wrap(localResult, *input, rows);
     }
+    context.moveOrCopyResult(localResult, rows, result);
```

Comment:
We may need to call moveOrCopyResult after saver is destroyed. See https://github.com/facebookincubator/velox/pull/2389

CC: @bikramSingh91

Reply:
I think the current code does call moveOrCopyResult after saver is destroyed. Not sure if I understand the concern. Could you clarify a bit?

Reply:
> I think the current code does call moveOrCopyResult after saver is destroyed.

I didn't notice this. Great. Thank you for pointing out.